### ISSUE TITLE
[patch] bug in bash-completion code 

### ISSUE THREAD

**Issue by biblicabeebli** [gitgitgadget/git#1064]
Hi, I wrote up this post based on the guidelines in Documentation/SubmittingPatches, and then was directed here as an alternative to `git send-email`.  I'm pasting below from the document I wrote up, I'm looking for feedback, though I believe my proposed patch is ... fine?  I have modified it slightly to improve the representation of the pasted content on github.

I have pasted this as an issue with a git patch contained within it.  If this should be reposted as a pull request, etc. please provide a little bit extra to that guideline so I can proceed correctly.

I am aware that the appropriate location of this patch submission may be the source of the bash-completion code external to git itself, this is part of the feedback I am looking for, I'm not sure where that canonical reference code lives.

Thanks for reviewing.

================

Hello!
This is my first attempt at providing a fix / bug notification with a proposed fix.  I'm doing my best to follow guidelines from Documentation/SubmittingPatches in a reasonable fashion.  I hope you will be patient with me, your requirements are quite extensive, and I'm unsure on the correct balance of correctness vs. reasonableness.

I have included my change as the output of a `git diff --patch` operaion at the end of this post.  The reason for this is that your first requirement, the base branch - I'm simply unsure what to provide here, I comment further below.

DESCRIPTION:
The git-completion.bash code may emit a warning directly onto the command prompt when attempting to use the autocompletion.  This user experience is terrible, it is obviously not intended, and renders the feature unusable.  This fix blocks that output and restores intended functionality.  This issue occurs for any bash alias that is operated on by the appropriate bash-completion function to enable the appropriate autocompletion.

Example: "some_file.py" is a random modified file in a random repository, it is the extent of expected output of the tab completion in this contrived case. `a` as an alias of `git add`, it has been operated on with `__git_complete a _git_add` in a bash profile file at shell-load-time.
```
$ a 
*user presses tab, updated terminal output is as follows*
$ a bash: [: -lt: unary operator expected
some_file.py
```

This behavior change (the `[: -lt: unary operator expected`) was obverved upon a standard update from Ubuntu 21.04 -> 21.10.  My working hypothesis is that this was a change to the `[` program, though it could also be a change in the Bash shell itself.  The fix is simple and in my opinion reasonable.  Bash experts may differ in their opinion, that is also reasonable.

Obscure, relevant Bash detail:
`[` is in fact a _program_; it mimics syntax, but is not syntax.  `[` has a very limited interface that does not appear to provide any mechanism for suppressing error or warning output.

Underlying cause:
In the script the variable `cword` is used as an argument to the `[` program in 6 places along with the `-lt` (less than) directive.  The `cword` variable is (apparently) potentially empty, and that results in a warning due to a missing parameter.

The fix proposed is to append `2> /dev/null` to all 6 instances of this call.  This is Bash-speak for directing all stderr output from `[` to /dev/null.  Voila, no more obscure error message upon pressing tab.  I acknowledge that the resultant bash code of the if-statement is ... weird.

System compatibility:
I know of no Unix or Unix-like system that lacks /dev/null, I am confident it is reasonable to expect it.  If there is some weird Bash emulation issue with the project over at gitforwindows.org, welp, its a good thing they are OSS.

BASE BRANCH:
I am unsure on the best thing to provide here.  The patch submission documentation provides guidelines for providing a patch, and literally my understanding of the technical terminology here is imperfect.

I am unsure what to do with output provided by running `git log --first-parent master..seen`, pretty sure it is a misunderstanding my part.  It provides me with merges from 3 days ago, October 18, 2021, which seems unlikely to be the intended result.

The only relevant git identifier I can provide is a commit hash based on the output of the following command, which maps to the six lines in my proposed change:
`git blame contrib/completion/git-completion.bash | grep "\-lt"`
The resultant commit hash is: da48616f1df
This is old, from 2010.  Consequently I just don't know how to address a base branch for this issue.

SIGN OFF:
(Not familiar with this / I don't have a commit for you? *insert shrug emoji here*)
I, Eli Jones <biblicabeebli@gmail.com>, hereby certify to the best of my knowledge and ability that this patch is in accordance with the Developer's Certificate of Origin 1.1.

```
diff --git a/contrib/completion/git-completion.bash b/contrib/completion/git-completion.bash
index eb5fd4783d..a57dcd8e89 100644
--- a/contrib/completion/git-completion.bash
+++ b/contrib/completion/git-completion.bash
@@ -1011,7 +1011,7 @@ __git_complete_remote_or_refspec ()
    if [ "$cmd" = "remote" ]; then
        ((c++))
    fi
-   while [ $c -lt $cword ]; do
+   while [ $c -lt $cword ] 2> /dev/null; do
        i="${words[c]}"
        case "$i" in
        --mirror) [ "$cmd" = "push" ] && no_complete_refspec=1 ;;
@@ -1187,7 +1187,7 @@ __git_find_on_cmdline ()
    done
    local wordlist="$1"
 
-   while [ $c -lt $cword ]; do
+   while [ $c -lt $cword ] 2> /dev/null; do
        for word in $wordlist; do
            if [ "$word" = "${words[c]}" ]; then
                if [ -n "${show_idx-}" ]; then
@@ -1283,7 +1283,7 @@ __git_get_option_value ()
 __git_has_doubledash ()
 {
    local c=1
-   while [ $c -lt $cword ]; do
+   while [ $c -lt $cword ] 2> /dev/null; do
        if [ "--" = "${words[c]}" ]; then
            return 0
        fi
@@ -1449,7 +1449,7 @@ _git_branch ()
 {
    local i c="$__git_cmd_idx" only_local_ref="n" has_r="n"
 
-   while [ $c -lt $cword ]; do
+   while [ $c -lt $cword ] 2> /dev/null; do
        i="${words[c]}"
        case "$i" in
        -d|-D|--delete|-m|-M|--move|-c|-C|--copy)
@@ -3213,7 +3213,7 @@ _git_svn ()
 _git_tag ()
 {
    local i c="$__git_cmd_idx" f=0
-   while [ $c -lt $cword ]; do
+   while [ $c -lt $cword ] 2> /dev/null; do
        i="${words[c]}"
        case "$i" in
        -d|--delete|-v|--verify)
@@ -3388,7 +3388,7 @@ __git_main ()
    local __git_C_args C_args_count=0
    local __git_cmd_idx
 
-   while [ $c -lt $cword ]; do
+   while [ $c -lt $cword ] 2> /dev/null; do
        i="${words[c]}"
        case "$i" in
        --git-dir=*)
```

**Comment by phil-blain** [comment]
Hi @biblicabeebli, to quote from [SubmittingPatches](https://git-scm.com/docs/SubmittingPatches#base-branch):

> - A bugfix should be based on `maint` in general. If the bug is not present in `maint`, base it on `master`. For a bug that’s not yet in `master`, find the topic that introduces the regression, and base your work on the tip of the topic.

From what I understand, your bugfix is for a bug you found in a released version of Git, so to choose your base branch, you could checkout `maint`, build Git and the Bash completion, install them,  and see if your bug reproduces. If so, you are done, you can use `maint` as your base branch. The above advice tells us that if the bug was _not_ reproducible on `maint`, but was reproducible on `master`, then base your topic branch on `master`.

In practice, however, most patch contributions, even bug fixes, are sent to the mailing list based on `master`, just as @dscho suggested. Then, when the maintainer, Junio, decides that the topic is ready for merging to `seen` first, then `next`, then `master`, he often changes the base commit himself to some earlier commit. But you do not have to care about that yourself most of the time.


The command `git log --first-parent master..seen` is suggested to find the hash of the tip of a topic branch (meaning, somebody else's contribution that's being considered for inclusion). The "Decide what to base your work on" section mentions that command because it's useful to know that hash if your bug fix is for a bug that you found on somebody else's topic, that is not yet integrated to  `master`. That's not the case here.

For the sign-off, it's sufficient to add the "Signed-off by:" trailer to your commit message, which can be done easily with `git commit -s`.

I hope that clears things up  :)

**Comment by biblicabeebli** [comment]
@phil-blain Thank you for the detailed comment!!

It's on me to move on this, I've found several more very similar instances of this bug on other autocompletions that I run less frequently.  I need to find the time to track them down.  A general solution may be something like redefining `[` in the script (eep!) to block error statements by default, and have a flag a dev can set to allow a more experienced dev to debug the source cause.

Maintainers may close this issue if that is reasonable on your end, otherwise I will return and comment if/when I get to this with my own fork and still need some help with getting it where it needs to go.

Thanks!

**Comment by biblicabeebli** [comment]
My issue appears to have been resolved in a future version of git....
https://askubuntu.com/questions/1371505/bash-completion-for-git-branches-broken-after-upgrading-to-21-10
https://github.com/git/git/pull/1038

**Comment by phil-blain** [comment]
Yes, in fact the fix is already in Git 2.33.0, as I just noted at the bottom of git#1038 :)
